index.js

```diff
@@ -1,301 +1,318 @@
-    .on( 'httpUploadProgress', function ( progress ) {
-        that.emit( 'uploading', path, s3.bucket + '/' + key, progress );
+    .then((gZipedPath) => {
+      return _upload(gZipedPath, stream)
     })
-    .on( 'success', function () {
-        that.emit( 'uploaded', path, s3.bucket + '/' + key );
-
-        // remove the current path from the queue
-        that._uploadq.shift();
-
-        // if it's not empty, continue to the next one
-        if ( that._uploadq.length ) {
-            that._upload();
-        }
-
-        // remove the uploaded file
-        that._unlink( path )
+    .tap((gZipedPath) => {
+      stream.emit('rotated', originalPath, gZipedPath || '')
     })
-    .send()
 }
 
-RotateStream.prototype._unlink = function ( path, cb ) {
-    var that = this;
-    this.emit( 'delete', path );
-    fs.unlink( path, function ( err ) {
-        if ( err ) {
-            that.emit( 'error', err );
-        } else {
-            that.emit( 'deleted', path );
-            if ( cb ) {
-                cb();
-            }
-        }
-    })
-}
-
-RotateStream.prototype._rotate = function ( suffix, cb ) {
-    var that = this;
-    var path = this.path + '.' + suffix;
-    var lock = this.path + '.lock';
-
-    // prevent multiple processes from rotating the same file
-    lockfile.lock( lock, { stale: 10000, wait: 5000 }, function ( err ) {
-        if ( err ) {
-            return _cb( err );
-        }
-
-        fs.stat( that.path, function ( err, stats ) {
-
-            // if no entry, file has been moved or removed - keep going
-            if ( err && err.code != 'ENOENT' ) {
-                return _cb( err );
-            }
-
-            // was the file externally renamed?
-            if ( !stats || that.ino != stats.ino ) {
-                return _cb();
-            }
-
-            that.emit( 'rotate', that.path, path );
-            fs.rename( that.path, path, function ( err ) {
-
-                // if no entry, file has been moved or removed - keep going
-                if ( err && err.code != 'ENOENT' ) {
-                    return _cb( err );
-                }
-
-                _cb();
-                that.emit( 'rotated', that.path, path );
-            })
+/**
+ * @private
+ */
+function _compress(path, stream) {
+  return new Promise((resolve, reject) => {
+    let pathGzip = path + '.gz'
+
+    if (!stream.compress) {
+      resolve(path)
+    } else {
+      stream.emit('compress', path, pathGzip)
+      fs.createReadStream(path)
+        .pipe(zlib.createGzip())
+        .pipe(fs.createWriteStream(pathGzip))
+        .once('error', (err) => {
+          reject(err)
         })
-    });
-
-    function _cb ( err ) {
-        lockfile.unlock( lock, function ( unlockerr ) {
-            cb( unlockerr || err );
+        .once('finish', () => {
+          stream.emit('compressed', path, pathGzip)
+          _unlink(path, stream)
+            .then(() => {
+              resolve(pathGzip)
```

Comment:
what about the rejection? you're quietly dismissing errors? will be cause the whole process to break because of uncaught exception?

Reply:
Added the missing 'catch' & 'reject'